### ISSUE TITLE
Future-proof Process structure

### ISSUE THREAD

**Issue by ODiogoSilva** [ODiogoSilva/assemblerflow#55]
## Context on assemblerflow's Processes

As detailed [in the development docs](http://assemblerflow.readthedocs.io/en/latest/create_process.html), every new component that we want to include in assemblerflow requires the creation of a new python class that inherits from the [Process](http://assemblerflow.readthedocs.io/en/latest/assemblerflow.generator.process.html#assemblerflow.generator.process.Process) abstract class. This `Process` class contains all the heavy-lifting code and attributes that allow assemblerflow's engine to build a pipeline and it abstracts the majority of the work from the person adding the new component. The new classes that are created, simply need to define attributes concerning the input and output types, parameters, secondary inputs, etc. So, adding a new process/module class can be as simple as:

```
class Skesa(Process):
    """Skesa process template interface"""

    def __init__(self, **kwargs):

        super().__init__(**kwargs)

        self.input_type = "fastq"
        self.output_type = "fasta"

        self.directives = {"skesa": {
            "cpus": 4,
            "memory": "{ 5.GB * task.attempt }",
            "container": "ummidock/skesa",
            "version": "0.2.0-3",
            "scratch": "true"
        }}
```

There are several other attributes that can be specified, but these classes are basically descriptions  of the component, with basically no additional code required.

## The problem

Currently, both the `Process`abstract class and all the actual classes that describe assemblerflow's modules are stored in the same file in `assemblerflow.generator.process`. However, this is hardly sustainable in the medium/long-term where more and more component classes are expected to be added. It also creates a problem for the automated generation of the documentation, since the API and description of the `Process` abstract class is mainly dev-oriented, while the descriptions of the actual components like `fastqc`, `spades, etc, are user-oriented. Finally, putting everything in the same file without any type of sorting or categorisation will make it hard to navigate this file.
If we want to make it easy for other people to contribute, this has to be improved.

## Potential solution

The most obvious solution is to move the component classes to another file, or directory in the assemblerflow package as leave the abstract `Process` class in the same place. Additionally, we could sort the classes by some categories, such as "quality control" or "assembly". So the idea would be to create a directory in

```
assemblerflow/generator/process_market # name up to discussion
```

And inside, create a file structure according to the type of component:

```
assemblerflow/generator/process_market/quality_control.py
assemblerflow/generator/process_market/assembly.py
(...)
```

The suggested categories for what we have now could be:

- `downloads.py`
- `quality_control.py`
- `assembly.py`
- `assembly_{polish|processing}.py`
- `annotation.py`
- `mlst.py`
- `mapping.py`

New categories could then be added as needed, we would just have to import the new modules in the engine. Another advantage of this approach would be the natural sorting of the classes into categories for the user documentation.

**Comment by tiagofilipe12** [comment]
I agree that this needs to be addressed. The separation of the "user-land" and "dev-land" is critical particularly because `Process` class is a "black box" for the users that want to add new processes. 

I think that your proposal is suitable. And if for some reason one of these `.py` files with processes grows too much we can always subdivide them in the future. Of course, that this would require editing the docs accordingly, but that seems a small price to such an improvement in the processes classes readability throughout assemblerflow.

**Comment by ODiogoSilva** [comment]
`annotation.py` would refer to functional annotation modules, such as `abricate` and `prokka`, while `mlst.py` would refer to typing methods. As you previously suggested, `mlst` could become `typing` to include `mlst`, `chewbbaca`, `patho_typing` and `seq_typing`?

As you noted, some processes may fit more than one category and in those cases I agree that having some simple advice guidelines to classify processes will be important to standardize the process.

**Comment by ODiogoSilva** [comment]
Closed via #65
